ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs

```diff
@@ -564,7 +598,48 @@ protocolInfoCardano protocolParamsByron@ProtocolParamsByron {
         reassoc ::
              NP (BlockForging m :.: ShelleyBlock) (ShelleyBasedEras c)
           -> OptNP 'False (BlockForging m) (CardanoEras c)
-        reassoc = OptSkip . injectShelley unComp . OptNP.fromNonEmptyNP
+        reassoc = OptSkip . injectShelleyOptNP unComp . OptNP.fromNonEmptyNP
+
+-- | When the given ledger state corresponds to a Shelley-based era, apply the
+-- given function to it.
+overShelleyBasedLedgerState ::
+     forall c.
+     (   forall era. EraCrypto era ~ c
+      => LedgerState (ShelleyBlock era)
+      -> LedgerState (ShelleyBlock era)
+     )
+  -> LedgerState (CardanoBlock c)
+  -> LedgerState (CardanoBlock c)
+overShelleyBasedLedgerState f (HardForkLedgerState (State.HardForkState st)) =
+    -- We can't use the pattern synonyms to match and /construct/ the various
+    -- 'LedgerState's because they are not bidirectional. Moreover, such a
+    -- solution would need to be extended for /each/ era. Instead, we do it
+    -- generically for all Shelley-based eras and then cons Byron in front.
+    HardForkLedgerState $ State.HardForkState $ hap fs st
+  where
+    fs :: NP (State.Current LedgerState -.-> State.Current LedgerState)
+             (CardanoEras c)
+    fs = fn id
+        :* injectShelleyNP
+             unwrapped
+             (hcpure (Proxy @(HasCrypto c)) (fn (Comp . f . unComp)))
+
+    unwrapped ::
+         (     LedgerState :.: ShelleyBlock
+          -.-> LedgerState :.: ShelleyBlock
+         ) shelleyEra
+      -> (     State.Current LedgerState
+          -.-> State.Current LedgerState
+         ) (ShelleyBlock shelleyEra)
+    unwrapped g = fn $ \(State.Current start state) ->
+          State.Current start
+        . unComp
+        . apFn g
+        . Comp
+        $ state
```

Comment:
Do the lifting of `State.Current` separately using https://github.com/input-output-hk/ouroboros-network/blob/376a5ba3a4e5a9c1dc68bdc1c64d875f3762abfe/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/State/Lift.hs#L18